lib/biokbase/catalog/db.py

```diff
@@ -120,130 +125,136 @@ class MongoCatalogDBI:
-                                        unique=True, sparse=False)
-
-        # hidden configuration parameters
-        self.secure_config_params.create_index('module_name_lc')
-        self.secure_config_params.create_index([
-            ('module_name_lc', ASCENDING),
-            ('version', ASCENDING),
-            ('param_name', ASCENDING)],
-            unique=True, sparse=False)
+                # Check db schema
+                if not self._db_schema_checked:
+                    self.check_db_schema()
+                    # Mark that we've checked the schema
+                    self._db_schema_checked = True
+
+            except ConnectionFailure as e:
+                error_msg = "Cannot connect to Mongo server\n"
+                error_msg += "ERROR -- {}:\n{}".format(
+                    e, "".join(traceback.format_exception(None, e, e.__traceback__))
+                )
+                raise ValueError(error_msg)
+
+    def _get_collection(self, collection_name):
+        """Lazily load collections."""
+        self._initialize_mongo_client()
+        if not self.index_created[collection_name]:
+            self._create_indexes(collection_name)
+            self.index_created[collection_name] = True
+        return self.db[collection_name]
+
+    # Define getters for each collection
+    @property
+    def modules(self):
+        return self._get_collection(MongoCatalogDBI._MODULES)
+
+    @property
+    def module_versions(self):
+        return self._get_collection(MongoCatalogDBI._MODULE_VERSIONS)
+
+    @property
+    def local_functions(self):
+        return self._get_collection(MongoCatalogDBI._LOCAL_FUNCTIONS)
+
+    @property
+    def developers(self):
+        return self._get_collection(MongoCatalogDBI._DEVELOPERS)
+
+    @property
+    def build_logs(self):
+        return self._get_collection(MongoCatalogDBI._BUILD_LOGS)
+
+    @property
+    def favorites(self):
+        return self._get_collection(MongoCatalogDBI._FAVORITES)
+
+    @property
+    def client_groups(self):
+        return self._get_collection(MongoCatalogDBI._CLIENT_GROUPS)
+
+    @property
+    def volume_mounts(self):
+        return self._get_collection(MongoCatalogDBI._VOLUME_MOUNTS)
+
+    @property
+    def exec_stats_raw(self):
+        return self._get_collection(MongoCatalogDBI._EXEC_STATS_RAW)
+
+    @property
+    def exec_stats_apps(self):
+        return self._get_collection(MongoCatalogDBI._EXEC_STATS_APPS)
+
+    @property
+    def exec_stats_users(self):
+        return self._get_collection(MongoCatalogDBI._EXEC_STATS_USERS)
+
+    @property
+    def secure_config_params(self):
+        return self._get_collection(MongoCatalogDBI._SECURE_CONFIG_PARAMS)
+
+    def _create_indexes(self, collection_name):
+        """Create indexes for the given collection lazily."""
+        collection = self.db[collection_name]
+
+        # Get the indexes for the collection from the DBIndexes class
```

Comment:
It seems that all of this code can go into the db_indexes.py code.

Reply:
It can. However, it makes more sense to me that in db_indexes.py, we first retrieve the indexes, and then the code to create the indexes is underneath the _create_indexes function.

Reply:
You mean line 242?

Reply:
If you mean the entire process, it makes more sense to me to separate the steps of getting and creating indexes.

Reply:
Does it look better now?

Reply:
no more `elif`. Call _create_indexes function in `__init__`

Reply:
I think the original way it looked was much better.

Reply:
We are reverting to the previous method because lazy Mongo load is no longer needed. It is done.

https://github.com/kbase/catalog/blob/dev-mongo7_upgrade/lib/biokbase/catalog/db.py#L210